pyccel/parser/scope.py

```diff
@@ -0,0 +1,325 @@
+
+        self._locals  = OrderedDict((k,OrderedDict()) for k in keys)
+
+        self._temporary_variables = []
+
+        if decorators:
+            self._locals['decorators'].update(decorators)
+
+        # TODO use another name for headers
+        #      => reserved keyword, or use __
+        self.parent_scope        = parent_scope
+        self._sons_scopes        = OrderedDict()
+
+
+        self._is_loop = is_loop
+        # scoping for loops
+        self._loops = []
+
+    def new_child_scope(self, name, **kwargs):
+        """
+        Create a new child Scope object which has the current object as parent.
+
+        The parent scope can access the child scope through the '_sons_scopes'
+        dictionary, using the provided name as key. Conversely, the child scope
+        can access the parent scope through the 'parent_scope' attribute.
+
+        Parameters
+        ----------
+        name : str
+            Name of the new scope, used as a key to retrieve the new scope.
+
+        kwargs : dict
+            Keyword arguments passed to __init__() for object initialization.
+
+        Returns
+        -------
+        child : Scope
+            New child scope, which has the current object as parent.
+
+        """
+
+        child = Scope(**kwargs, parent_scope = self)
+
+        self._sons_scopes[name] = child
+        child.parent_scope = self
+
+        return child
+
+    @property
+    def imports(self):
+        """ A dictionary of objects imported in this scope
+        """
+        return self._imports
+
+    @property
+    def variables(self):
+        """ A dictionary of variables defined in this scope
+        """
+        return self._locals['variables']
+
+    @property
+    def classes(self):
+        """ A dictionary of classes defined in this scope
+        """
+        return self._locals['classes']
+
+    @property
+    def functions(self):
+        """ A dictionary of functions defined in this scope
+        """
+        return self._locals['functions']
+
+    @property
+    def macros(self):
+        """ A dictionary of macros defined in this scope
+        """
+        return self._locals['macros']
+
+    @property
+    def headers(self):
+        """A dictionary of user defined headers which may
+        be applied to functions in this scope"""
+        return self._locals['headers']
+
+    @property
+    def templates(self):
+        """A dictionary of user defined templates which may
+        be applied to functions in this scope"""
+        return self._locals['templates']
+
+    @property
+    def decorators(self):
+        """Dictionary of Pyccel decorators which may be
+        applied to a function definition in this scope."""
+        return self._locals['decorators']
+
+    @property
+    def static_functions(self):
+        """ A dictionary of static functions defined in this scope
+        """
+        return self._locals['static_functions']
```

Comment:
@saidctb Do you know what a static function is?

Reply:
it's probably not needed anymore

Reply:
Fixed: 80886bf  af002f6